### PR TITLE
Add internal locking to AudioQueue

### LINKED ISSUE #1458: Xournal++ crashes when audio finishes playing
**Affects versions :**
 - OS:  Fedora 30 Gnome Wayland And X11(occurs on both), Built and compiled
 - Version of Xournal++ : Most recent one, seems to be master on commi 6260dba

**Describe the bug**
After recording, when I click Play object, once the audio ends, it crashes horribly, in one case froze on the GUI and needed to restart PC.

**To Reproduce**
1. Step 1: Build and Compile xournalpp on Fedora 30, package as a .tar.gz
2. Step 2: Set audio input and output to pulse, Record, and then play audio.
3. Step 3: Once audio finishes playing, it will crash

**Expected behavior**
Not to crash
[errorlog.20190906-231602.log](https://github.com/xournalpp/xournalpp/files/3586213/errorlog.20190906-231602.log)

EDIT: To clarify about what I meant when I said: once the audio ends: It seems to occur both when general audio finishes playing, as well as when audio finishes playing for a particular object

### PR DESCRIPTION
**Pull request by LittleHuba**
@morrolinux Please test this on your machine

Fixes #1458

### CODE REVIEW

File: src/util/audio/AudioQueue.h [hunk 1]
```diff
@@ -100,7 +100,7 @@ class AudioQueue : protected std::deque<T>
 
 		{
 			std::lock_guard<std::mutex> lock(internalLock);
-			returnBufferLength = std::min(nSamples, std::deque<T>::size() - std::deque<T>::size() % this->channels);
+			returnBufferLength = std::min(nSamples, (unsigned long)(std::deque<T>::size() - std::deque<T>::size() % this->channels));
```
Comment (Febbe) [inline src/util/audio/AudioQueue.h]: `returnBufferLength = std::min<unsigned>(nSamples, std::deque<T>::size() - std::deque<T>::size() % this->channels));` should work also. 
But I would rather change returnBufferLength to size_t.

File: src/util/audio/AudioQueue.h [hunk 2]
```diff
@@ -91,11 +97,14 @@ class AudioQueue : protected std::deque<T>
...
 		}
 
-		returnBufferLength = std::min(nSamples, this->size() - this->size() % this->channels);
-		for (long i = 0; i < returnBufferLength; i++)
 		{
-			returnBuffer[i] = this->back();
-			this->pop_back();
+			std::lock_guard<std::mutex> lock(internalLock);
+			returnBufferLength = std::min(nSamples, std::deque<T>::size() - std::deque<T>::size() % this->channels);
+			for (unsigned long i = 0; i < returnBufferLength; i++)
```
Comment (Febbe) [inline src/util/audio/AudioQueue.h]: use size_t

File: src/util/audio/AudioQueue.h [hunk 3]
```diff
@@ -52,23 +53,28 @@ class AudioQueue : protected std::deque<T>
...
 
 	void push(T* samples, unsigned long nSamples)
 	{
 		XOJ_CHECK_TYPE(AudioQueue);
 
-		for (unsigned long i = 0; i < nSamples; i++)
 		{
-			this->push_front(samples[i]);
+			std::lock_guard<std::mutex> lock(internalLock);
+			for (unsigned long i = 0; i < nSamples; i++)
```
Comment (Febbe) [inline src/util/audio/AudioQueue.h]: use size_t

File: src/util/audio/AudioQueue.h [hunk 4]
```diff
@@ -52,23 +53,28 @@ class AudioQueue : protected std::deque<T>
 	{
 		XOJ_CHECK_TYPE(AudioQueue);
 
+		std::lock_guard<std::mutex> lock(internalLock);
 		return std::deque<T>::empty();
 	}
 
 	unsigned long size()
```
Comment (Febbe) [inline src/util/audio/AudioQueue.h]: return size_t

File: src/util/audio/AudioQueue.h [hunk 5]
```diff
@@ -16,6 +16,7 @@
 #include <vector>
 #include <deque>
 #include <mutex>
+#include <cmath>
```
Comment (Febbe) [inline src/util/audio/AudioQueue.h]: is this for std::min? then I would include `<algorithm>`

**Comment by Febbe** [comment]
I think it would be better, not to return internal states of the container like size() or empty(). For example the internal lock does not prevent the container size to change between a call to size() and pop(). I would only use functions like push, pop, try_pop, for_each, consume, etc.

**Comment by LittleHuba** [comment]
> I think it would be better, not to return internal states of the container like size() or empty().

I'm actually against this for a good reason. These methods are inherently save by the way they are used in the code and the code is structured. 

We have a setup of exactly one producer and one consumer. 

`size`:
- If we use `size` on the producer side, we will check for an overflow. The consumer may only consume elements thus making the situation better for us even when the state is then not absolutely consistent. 
- If we check on the consumer side we check for an underflow which is either the case and we abort whatever we do or if not the producer may push new frames inbetween calling `size` and `pop` not changing the situation for the consumer.

empty:
- Using `empty` on the consumer side, we check if there currently are no samples available which might be used to pause the polling until samples are available again. When the producer pushes samples he will unblock the consumer, thus ensuring that it polls again. Should this not succeed since the notification arives between the call to `empty` and the call to `wait` the either the next push or the call to `signalEndOfStream` wakes the consumer up again.
- Using `empty` on the producer side would mean we want to make sure that the queue is empty before pushing anything to it which makes absolutely no sense at all. Even if we do this, again the same rule applies. The consumer may only consume elements, making the condition true even when we got a wrong state, again notifying the producer if it polls... I think you get what I mean.

So long story short: Changing this is only overhead and has no advantage.

**Comment by morrolinux** [comment]
> @morrolinux Please test this on your machine
> 
> Fixes #1458

I'm sorry, for some odd reasons which I still have to figure out, I don't receive github emails anymore and therefore lose notifications!!
I'm gonna try this ASAP

**Comment by morrolinux** [comment]
@LittleHuba Tested and works allright :)
